README.md
# service.subtitles.traduttorianonimi
[Kodi Addon] Servizio Sottotitoli per [www.traduttorianonimi.it](http://www.traduttorianonimi.it/)

## :it:Aggiungere l'italiano alle "Lingue per cui scaricare i sottotitoli"<br>:us:How to add italian to "Languages to download subtitles for"
:it:Sistema -> Video -> Sottotitoli --> Lingue per cui scaricare i sottotitoli --> Aggiungi Italiano<br>
:us:System -> Video -> Subtitles -> Languages to download subtitles for -> Add Italian<br />

## :it:Come Installare<br>:us:How to Install<br>

### Repository
:it:Sistema -> Add-on-> Installa da repository  -> Kodi Add-on repository -> Servizio Sottotitoli -> traduttorianonimi.it<br>
:us:System-> Add-ons -> Install from repository -> Kodi Add-on repository -> Subtitles -> traduttorianonimi.it<br>

### ZIP
- Scarica il file Zip più recente / Download lastest Zip file: [Releases](https://github.com/ShellAddicted/service.subtitles.traduttorianonimi/releases)
- Segui Queste istruzioni / Follow this instructions:<br />[Kodi Wiki](http://kodi.wiki/view/HOW-TO:Install_add-ons_from_zip_files)

## Todo
- [ ] Supporto ad i Film / Support for Movies
- [ ] Controllo Versione dei sottotitoli (Sync) / Check Subtitles Version (Sync)




changelog.txt
1.0.0
- Initial Release

2.0.0
- Features -> Support for new website http://www.traduttorianonimi.it/ instead of http://www.traduttorianonimi.webbly.com/ 
- Code:
-  Source Code Cleanup
-  Now using bs4 from shared module script.module.beautifulsoup4

2.0.1
- Feautures -> Nothing
- Code:
-  variable name style like "__addon__" replaced with "addon"
-  added tag source in addon.xml

2.0.2
- Feautures -> Nothing
- Code:
-  Support for unicode(utf-8) TV Shows Name
-  deleted duplicated functions and variables from service.py and traduttorianonimi.py, now they are shared by util.py
-  Localized Notfications (Italian,English)

2.0.3
- Feautures -> Nothing
- Code:
    Converted strings.xml to strings.po

2.0.4
- Feautures -> Nothing
- Code:
    Fixed errors for android devices

2.1.0
- Feautures -> Partial Sync check (more work to do)
- Code:
    Code Cleanup

2.1.1
- BugFix

2.2.0
- Features:
    Removed Sync check due to reported issues
    parser is now fully compatible with new website structure
- Code:
    Code Cleanup


resources/lib/TraduttoriAnonimi.py
# -*- coding: utf-8 -*-
#
#  TraduttoriAnonimi.py
#  
#  Copyright 2016 ShellAddicted <shelladdicted@gmail.com<>
#  
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 3 of the License
#  
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.
#  
#
import xbmc
import xbmcgui
import xbmcaddon    
from bs4 import BeautifulSoup
import requests
import re
import sys
import urlparse
import os

addon = xbmcaddon.Addon()
cwd = xbmc.translatePath(addon.getAddonInfo("path")).decode("utf-8")
resource = xbmc.translatePath(os.path.join(cwd, "resources", "lib")).decode("utf-8")

sys.path.append(resource)
from utils import *
log=log #.getChild("Core") #Disabled for android compatibility

def MagicInt(x): #Cast to integer (if possible) without Errors
    try:
        return int(x)
    except:
        return x

class SearchableDict(dict):
    def __getitem__(self, token):
        p=re.compile(MagicUnicode(token),re.IGNORECASE)
        result=[]
        for x,y in self.items():
            if re.search(p,MagicUnicode(x)):
                result.append({"Name":x,"URL":y})
                
        if (len(result)==0):
            result=None
        
        return result
    def __setitem__(self,i,y):
        i=MagicUnicode(i)
        y=MagicUnicode(y)
        return dict.__setitem__(self, i,y)

class TraduttoriAnonimi:
    def __init__(self,BaseURL="http://traduttorianonimi.it"):
        self.log=log #.getChild("TA") #Disabled for android compatibility
        
        self.BaseURL=BaseURL
        self.log.debug("BaseURL=> {0}".format(self.BaseURL))
        
        self.ShowsListURL=urlparse.urljoin(self.BaseURL,"elenco-serie/")
        self.log.debug("ShowsListURL=> {0}".format(self.ShowsListURL))
        
        self.ShowsList=SearchableDict()
        self.UpdateShowsList()
        
    def UpdateShowsList(self):
        self._GrabShowsListFromWebsite()
        
    def _GrabShowsListFromWebsite(self):
        self.log.info("Grabbing Shows list from Website.")
        url=self.ShowsListURL
        self.log.debug("self.ShowsList re-initialized")
        self.ShowsList=SearchableDict()
        for attempt in xrange(100):
            r=RetriveURL(url)
            self.log.debug("Attempt #{0}".format(attempt))
            if (r!=None):
                html=r.content
                parser=BeautifulSoup(html,"html.parser")
                
                for ShowTag in parser.findAll("a",{"href":re.compile("serie\?c=\d+", re.IGNORECASE)}):
                    ShowName=ShowTag.find("img").attrs["title"]
                    self.ShowsList[MagicUnicode(ShowName)]=ShowTag.attrs["href"]
                
                tmp=parser.find("div",{"class":"pagination" }).find("a",{"class":"next"})
                if (tmp!=None):
                    url=urlparse.urljoin(self.BaseURL,tmp.attrs["href"])
                    self.log.debug("Next Page Found. Following => {0}".format(url))
                else:
                    self.log.debug("Next Page Not Found. Exiting from loop")
                    break
            else:
                log.error("respose is None. Exiting from loop")
                break
        return self.ShowsList
    
    def GrabSubtitle(self,ShowName,Season,Episode):
        ShowName=MagicUnicode(ShowName)
        ShowsResults=self.ShowsList[ShowName]
        EpisodeRegex=re.compile("(?P<tvshowname>.+)(?:(?:\s|s|\.)|\.s|\.so)(?P<season>\d+)(?:x|e|\.x|\.e)(?P<episode>\d+)", re.IGNORECASE)
        if (ShowsResults == None):
            self.log.info("No ShowsResults")
            return []
        for show in ShowsResults:
            r=RetriveURL(show["URL"])
            if (r!=None):
                html=r.content
                parser=BeautifulSoup(html,"html.parser")
                        
                for ep in parser.findAll("td",{"class":"dwn"}):
                    tmp=ep.find("a")
                    self.log.debug("Analyzing => {0}".format(tmp))
                    if (tmp!=None and "title" in tmp.attrs and "href" in tmp.attrs):
                        x=EpisodeRegex.search(tmp.attrs["title"])
                        self.log.debug("Regex Groups=> {0}".format(x.groups()))
                        if (MagicInt(x.group("season"))==Season and MagicInt(x.group("episode"))==Episode):
                            self.log.info("Subtitle Found")
                            return [{"Name":tmp.attrs["title"],"URL":tmp.attrs["href"]}]



service.py
# -*- coding: utf-8 -*-
#  Copyright 2018 ShellAddicted <shelladdicted@gmail.com>
#  
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 3 of the License
#  
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.
#  
__author__ = "ShellAddicted"
__copyright__ = "Copyright 2018, ShellAddicted"
__license__ = "GPL"
__version__ = "2.2.0"
__maintainer__ = "ShellAddicted"
__email__ = "shelladdicted@gmail.com"
__status__ = "Development"

import xbmc
import xbmcvfs
import xbmcaddon
import xbmcgui
import xbmcplugin

import os
import sys
import logging

try:
    from urllib import unquote
except (ImportError, AttributeError):
    from urllib.parse import unquote

try:
    import StringIO
except ImportError:
    from io import StringIO

import requests
import zipfile
import shutil

addon = xbmcaddon.Addon()
cwd = xbmc.translatePath(addon.getAddonInfo("path")).decode("utf-8")
resource = xbmc.translatePath(os.path.join(cwd, "resources", "lib")).decode("utf-8")
profile = xbmc.translatePath(addon.getAddonInfo("profile")).decode("utf-8")
temp = xbmc.translatePath(os.path.join(profile, "temp", "")).decode("utf-8")
dialog = xbmcgui.Dialog()
author = addon.getAddonInfo("author")
scriptid = addon.getAddonInfo("id")
scriptname = addon.getAddonInfo("name")
version = addon.getAddonInfo("version")
language = addon.getLocalizedString
main_url = "http://traduttorianonimi.it"

NotifyLogo = xbmc.translatePath(os.path.join(cwd, "resources", "likelogo.png")).decode("utf-8")

if not os.path.isfile(NotifyLogo):
    NotifyLogo = None

def notify(message, header=scriptname, icon=NotifyLogo, time=5000):
    dialog.notification(header, message, icon=icon, time=time)

class LogStream(object):
    def write(self, data):
        xbmc.log("*** [service.subtiles.traduttorianonimi] -> {0}".format(data.encode('utf-8', "ignore")),
                 level=xbmc.LOGNOTICE)

logging.basicConfig(
    level=logging.INFO,
    format="{%(levelname)s} %(name)s.%(funcName)s() -->> %(message)s",
    handlers=[
        logging.StreamHandler(LogStream())
    ]
)
log = logging.getLogger("TraduttoriAnonimiKodiService")

sys.path.append(resource)
import TraduttoriAnonimi

core = TraduttoriAnonimi.TraduttoriAnonimi()

def cleanup():
    # cleanup junk files
    for x in [profile, temp]:
        if xbmcvfs.exists(x):
            shutil.rmtree(x)
        xbmcvfs.mkdirs(x)

def getParams():
    param = {}
    if len(sys.argv[2]) >= 2:
        for pairsofparam in sys.argv[2].replace("?", "").split("&"):
            tmp = pairsofparam.split("=")
            if len(tmp) == 2:
                param[tmp[0]] = tmp[1]
    return param

def checkSync(fnvideo, fnsub):
    #TODO: implement this
    return True

def search(item):
    cleanup()
    if "ita" in item["languages"]:
        if item["tvshow"]:
            results = core.getSubtitles(item["tvshow"], int(item["season"]), int(item["episode"]))
            if results is not None:
                for result in results:
                    subs = download(result["URL"])
                    for sub in subs:
                        # sub["Name"]
                        listitem = xbmcgui.ListItem(label='Italian', label2=os.path.basename(sub),thumbnailImage='it')
                        listitem.setProperty('sync', 'true' if checkSync(item["file_original_path"],os.path.basename(sub)) else 'false')
                        xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]),
                                                    url="plugin://{0}/?action=download&url={1}".format(scriptid, sub),
                                                    listitem=listitem, isFolder=False)  # sub["URL"]
                xbmcplugin.endOfDirectory(int(sys.argv[1]))  # send end of directory to XBMC
            else:
                log.info("NO RESULTS")
        else:
            # The Subtitles are available only for TV Shows, MOVIES will be supported in future
            notify(language(30002), time=18000)
            log.info('TraduttoriAnonimi only works with tv shows. Skipped')
    else:
        # The Subtitles are available only in Italian,Check Your Kodi Settings
        # [Video->Subtitles->Languages to download subtitles for]
        notify(language(30003), time=31000)
        log.info('TraduttoriAnonimi only works with italian. Skipped')


def download(url):
    if os.path.isfile(url):
        return [url]
    exts = ["srt"]
    out = []

    r = core._retriveURL(url)
    if r is not None:
        content = r.content
        tmp = StringIO.StringIO(content)  # "with" can't be used because StringIO has not __exit__
        if content[0] == "P":
            log.info("ZipFile Detected")
            q = zipfile.ZipFile(tmp)  # "with" can't be used because zipfile has not __exit__
            for name in q.namelist():
                if name.split(".")[-1] in exts:
                    q.extract(name, temp)
                    out.append(os.path.join(temp, name))
        else:
            log.info("Unpacked file detected")
            if os.path.basename(url).split(".")[-1] in exts:
                with open(os.path.join(temp, os.path.basename(url)), "wb") as q:
                    q.write(content)
                out.append(os.path.join(temp, os.path.basename(url)))
            else:
                log.info("Downloaded File ({0}) is not in exts[{1}]".format(os.path.basename(url), str(exts)))
    return out


def main():
    log.info("Traduttori Anonimi Kodi Service version: {0}".format(version))
    if xbmc.Player().isPlayingVideo():
        params = getParams()

        if params["action"] == "search" or params["action"] == "manualsearch":
            item = {}

            if params["action"] == "manualsearch": #ignore manual search is not supported
                notify(language(30004), time=5000)  # Manual Search is not supported yet, but it will be in the future.

            item["season"] = str(xbmc.getInfoLabel("VideoPlayer.Season"))
            item["episode"] = str(xbmc.getInfoLabel("VideoPlayer.Episode")).zfill(2)
            item["tvshow"] = xbmc.getInfoLabel("VideoPlayer.TVshowtitle")
            item["file_original_path"] = xbmc.Player().getPlayingFile().decode("utf-8")
            item["languages"] = []

            for lang in unquote(params["languages"]).decode("utf-8").split(","):
                item["languages"].append(xbmc.convertLanguage(lang, xbmc.ISO_639_2))

            if item["episode"].lower().find("s") > -1:  # Check if season is "Special"
                item["season"] = "0"
                item["episode"] = item["episode"][-1:]
            search(item)

        elif params["action"] == "download":
            subs = download(params["url"])
            for sub in subs:
                listitem = xbmcgui.ListItem(label=sub)
                xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=sub, listitem=listitem, isFolder=False)
            xbmcplugin.endOfDirectory(int(sys.argv[1]))

if __name__ == "__main__":
    main()
